### ISSUE TITLE
eigenvector_centrality: another wrong result

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1474]
Consider this graph:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/1212871/94292900-da6f7380-ff5d-11ea-9e2d-c4987dc9f4fa.png">

The correct eigenvector centralities are 

```
{1., 0.481194, 0.193937}
```

This is what I get in the unweighted case. However, if I assign weight 1 to each edge, I get:

```
{1., 0.249141, 0.0586332}
```

This is not correct.

The reason for this behaviour is that `igraph_inclist_remove_duplicate` fails to remove duplicate edges in this case. It assumes that the inclist is sorted and it removed only consecutive duplicates. However, this inclist is not sorted. It looks like this:

```
1 0 1 0 2 
2 3 
3 
```

`1` and `0` are the edge indices of the two self-loops.

**Comment by szhorvat** [comment]
I can't come up with a _good_ solution to this, except to add a flag to `igraph_incident` to ask it not to return self-loops twice (i.e. Vincent's suggestion). We could do that as an internal function for 0.8.3.

Any opinions? Easier/quicker fixes for 0.8.3?

**Comment by vtraag** [comment]
We cannot change `igraph_incident` for 0.8.3 of course because it breaks API.

We can easily remove duplicates by using a fixed set (i.e. a Boolean vector to keep track of whether a neighbour is already included). Removing self-loops explicitly can of course also be done in a similar fashion, but even simpler. This could be implemented in `remove_duplicates`.

**Comment by szhorvat** [comment]
> We cannot change `igraph_incident` for 0.8.3 of course because it breaks API.

No, we can't, but we could have a different, internal version.

```c
int igraph_i_incident(
                    const igraph_t *graph, igraph_vector_t *eids,
                    igraph_integer_t pnode, igraph_neimode_t mode,
                    igraph_bool_t duplicate_loops) { ... }

int igraph_incident(
                    const igraph_t *graph, igraph_vector_t *eids,
                    igraph_integer_t pnode, igraph_neimode_t mode) {
    return igraph_i_incident(graph, eids, pnode, mode, 1 /* true */);
}
```

> We can easily remove duplicates by using a fixed set (i.e. a Boolean vector to keep track of whether a neighbour is already included).

I thought of that, but I was a bit concerned about memory use. The memory requirements would be comparable to that of the graph. Perhaps that's not that bad though. It's certainly easier to implement that the other option.

**Comment by vtraag** [comment]
> The memory requirements would be comparable to that of the graph. 

It would be only O(n) right? So that is not too bad. However, we wouldn't want to allocate it anew every time, so we should need some "scratchpad" structure for this I think.

I think this issue could also use more discussion in the context of revising the internal graph data structure, see our [discussion](https://igraph.discourse.group/t/new-graph-data-structure-proposal/420/9) on the Discourse forum.

**Comment by szhorvat** [comment]
O(n) where n is the edge count. We can make it a `char` vector to save some space. 

@ntamas Any input on your preferences about how to fix this? If not, I'll just do it with the Boolean vector approach for simplicity.

**Comment by vtraag** [comment]
No, O(n) where n is the number of nodes. For each node we only need to consider whether its neighbours are already included in the return set. That means that we need to consider for n nodes whether they are already included.

Or should it remove only duplicate edges really (i.e. same edge index)? This only ever happens with self-loops, no? In that case couldn't we simply gather self-loops in a separate vector, and check that vector to see if that exact edge is already present? That takes a bit more time, but it would save a lot of space indeed, because we would then need O(m) space, with m the number of edges. The number of cases where this happens is rare anyway I think, so the performance decrease should be limited.

From recollection edge indices are not sorted anyway, it is only the neighbours that are sorted, i.e. edge indices are sorted in order of their neighbours. Or did I remember this incorrectly?

Best yet would be to include self-loops only once in `igraph_incident`, definitely from the performance point of view. But this is 0.9 only of course.

**Comment by vtraag** [comment]
Anyway, the internal alternative sounds most attractive for 0.8.3, that way we wouldn't need to remove duplicates. We could make that public then for 0.9.

**Comment by vtraag** [comment]
I was planning on pushing a solution with an internal function `igraph_i_incident` that accepts a `loop` argument to indicate whether you want no loops, 1 loop or 2 loops (similar to what was suggested in #1193). 

However, this raises a number of questions.

First of all, it raises the question how to deal with `inclist_remove_duplicates`. At the moment, this function does not work correctly because the edges are not sorted in order of the edge index, but in order of their neighbors. Of course, an edge index is really only identical when the exact edge is identical. Another question is whether this functionality should remove duplicate edge indices or actually parallel edges (i.e. those that have the same `from` and `to`).

The same question holds for `adjlist_remove_duplicate`. This does work correctly for `IGRAPH_IN` and `IGRAPH_OUT` arguments, but not for `IGRAPH_ALL`, since it will contain the concatenation of two sorted lists, which is not sorted of course. If we want it to work correctly we should merge the two sorted lists. It then begs the question of the exact order in `igraph_neighbors`, which should correspond to the order of `igraph_incident`. Moreover, `adjlist_remove_duplicates` does remove duplicate neighbors, and it hence does something else than `inclist_remove_duplicates` (at least at the moment).

Taking all of this in consideration, I would like to suggest the following. We should remove the `*remove_duplicate` functions and simply always include only a single loop in `igraph_incident` and in `igraph_neighbors`. That way, the definition is straightforward, simple to use and unambiguous. If a certain application wants to count self-loops twice (or not at all) they can easily do this in their own code. We should still make a choice about whether `igraph_incident` and `igraph_neighbors` should actually return sorted vectors or not. This is only relevant for the `IGRAPH_ALL` argument.

The `*remove_duplicate` functions are not a proper design choice I think. If its intention is to remove parallel edges, I think this should be left to the user, who has to do this via `igraph_simplify`. Otherwise, things will become messy around questions about how to deal with edge weights and other associated definitions. If its intention is to only remove self-loops then (a) the functions should be named `*remove_loops` and (b) this adds little benefit over simply skipping self-loops. All of this should be left for a 0.9 version because it will break API (and existing functionality without breaking API, perhaps an even more serious problem).

So, what can we do for 0.8.3? We simply remove the `remove_duplicate` call here

https://github.com/igraph/igraph/blob/f88f9c1aa4843a33b5380e06dcfcf0e214ee4ee9/src/centrality.c#L255

After all, it does not function correctly, and I don't see the problem when we remove this line. Or does this still leave the original problem unaddressed that the weighted version yields a different answer than the non-weighted response?

**Comment by ntamas** [comment]
I totally agree with the diagnosis and that we should get rid of the `remove_duplicate()` function in later versions. However, that's for later and my focus right now is on getting this fixed for 0.8.3.

For igraph 0.8.3, removing the call you mentioned might or might not be sufficient, depending on how you interpret the eigenvector centrality on unweighted graphs with self-loops :) The problem is that we loop through the incidence lists in `igraph_i_eigenvector_centrality2()` and perform a matrix-vector multiplication, increasing `to[i]` (the i-th component of the result vector of the matrix-vector multiplication) with `w * from[j]` for every edge going from j to i, where _w_ is the weight of said edge. For self-loops, you will essentially increase `to[i]` with `2 * w * from[i]` because the edge leading from `i` to `i` appears twice in the incidence list. If you think about the eigenvector centrality as a measure that is formulated from a random walk, then this is correct, because you are twice as more likely to follow a self-loop from node _i_ than any other edge incident on node _i_ since a self-loop has two ends incident on _i_.

If removing the call brings the weighted case in line with the unweighted case, then it's okay for me. I'll give it a go and let you know how it went.

**Comment by vtraag** [comment]
@szhorvat and I briefly discussed. It seems that this line is also trying to remove duplicate loops for the unweighted case:

https://github.com/igraph/igraph/blob/f88f9c1aa4843a33b5380e06dcfcf0e214ee4ee9/src/centrality.c#L239

I am not 100% sure whether `igraph_i_eigenvector_centrality_loop` is actually correctly implemented, I have not yet checked it. 

There would be three possible solutions:

1) We let this bug in for 0.8.3 and only solve it in 0.9.0
2) We fix it by implementing correct (and possibly slow) functions for removing duplicate loops for both weighted and unweighted case.
3) We remove both lines that are trying to remove duplicate loops.

**Comment by ntamas** [comment]
I guess first we would then need to decide how to handle loop edges with eigenvector centrality in general, i.e. what the correct solution should be. The possibilities are:

* completely ignore loop edges (as if they were not there at all)
* treat loop edges according to the random walk formulation (i.e. they are twice as likely to be traversed) so their weight counts twice
* treat loop edges such that their weight counts only once

A nice-to-have property would be that the eigenvector centrality of all nodes in an undirected graph should be the same as in its directed equivalent where each edge is replaced with a mutual edge pair, which rules out the last option (in my opinion).

**Comment by vtraag** [comment]
From practical consideration, the second option (treating them twice, counting them doubly) is perhaps also the most attractive for now in 0.8.3, as it would just mean removing the two lines.

In future versions we could pass arguments to indicate whether no loops should be considered, 1 loop or 2 loops.

**Comment by szhorvat** [comment]
So now effectively we have 2 on the diagonal of the adj. matrix for each self loop. This should be documented too.
